### ISSUE TITLE
stats sink push when envoy is not fully initialized

### ISSUE THREAD

**Issue by ramaraochavali** [envoyproxy/envoy#4420]
*Description*:
In one of our prod instances, we found that because of an issue with our management server, we did not provide ClusterLoadAssignment response to an EDS cluster. So that EDSCluster's `onPreInit` is not called and Envoy is still waiting at initializing secondary clusters and listeners are not fully initialized.
At this point, we do not get stats sink push also. What do you think about pushing stats data via stats sinks even though Envoy is not fully initialized? Is there a merit in doing this?

**Comment by ramaraochavali** [comment]
Update: In this mode when it is waiting on secondary clusters and if not all secondary clusters EDS updates are given, it just flushes the stats once and does not repeat it at `stats_flush_interval` which is kind of surprising

**Comment by htuch** [comment]
@ramaraochavali this sounds like a bug. The main thread should not be blocked and should be capable of periodic scheduling of stats flushes. Do you want to investigate?

**Comment by ramaraochavali** [comment]
Yup. That's what I thought. I can investigate but not immediately though. Will update what I find.

**Comment by ramaraochavali** [comment]
@mattklein123 looked at it. I think it looks like a chicken and egg problem. During server initialization, as part of config initialization, cluster manager is initialized before stats sinks here

https://github.com/envoyproxy/envoy/blob/master/source/server/configuration_impl.cc#L57

I tried to make `initializeStatsSinks` public and initialize it before cluster manager is initialized so that stats would be flushed even if cluster manager is stuck for some reason. But some of the stats sinks (especially metric service) is dependent on cluster manager initialization for getting the Async client factory.

Not sure if there is any better way to initialize sinks before cluster manager with out any side affects and also not sure if it is worth pursuing this further.

If it is not worth it, I am inclined to close this. WDYT?

**Comment by mattklein123** [comment]
I think it's probably doable in certain cases (like if stats is using static cluster config for example), but yeah, in more complex cases it might not be possible. I agree it's probably not worth it and I would probably close if you are OK with that.

**Comment by ramaraochavali** [comment]
@mattklein123  Yes. I am inclined to close this. Please close this.
